enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,668 @@
+
+`openshift-tests` will plan an efficient testing strategy based on the 
+extension binaries it discovers, the configuration profiles they support,
+and the tests that need to run in each configuration. It will:
+1. Discover extension binaries.
+2. Discover extension binary metadata (`info` verb), including configuration profiles.
+3. Discover tests that should be run for each configuration profile (`list` verb with `--config <profile>`, for each profile).
+4. Plan test execution by grouping configuration profiles into buckets.
+
+After `openshift-tests` plans test orchestration, it will proceed to execute tests in
+each configuration bucket. For a given bucket, `openshift-tests`
+will:
+1. Iterate through all involved components and configure them with non-conflicting configuration profiles (`config` verb).
+2. Await a high-level cluster steady state (cluster operators available, machines config up-to-date) if any configuration profile requires it.
+3. Run tests within the bucket with non-conflicting, test-level parallelization.
+
+Between buckets, the component may be asked to set its required `default` profile in order
+to return to its install-time configuration.
+
+#### Test Result Aggregation
+
+`openshift-tests` will store test results in the CI artifacts repository in a JSONL file.
+The output will be similar to that of the `run-test` output, but each test result 
+will be expanded to include:
+- Component & subcomponent
+- Environment
+- Testing Context
+
+The goal here is to allow a single file to contain comprehensive information necessary for
+a system like Component Readiness to analyze it without, for example, needing to draw information
+from prowjob names.
+
+```python
+{
+    # Human-readable test name
+    "name": "openssl security compliance",
+    
+    # Original human-readable test name, if it has changed over time.
+    "originalName": "fips security compliance",
+    
+    # The Test ID openshift/origin has defined for this test based
+    # on component and test metadata. It is meant to be unique
+    # across all other components and consistent across time
+    # (even if the human-readable name for a unit of test logic
+    # is changed).
+    "id": "openshift-payload-api-server-fips security compliance",
+    
+    "result": "pass",
+
+    # ...other test result information...
+    
+    # Information necessary to fully qualify the 
+    # test.
+    "component": {
+        "product": "openshift",
+        "type": "payload",
+        "component": "hyperkube",
+        "subcomponent": "logs",
+    },
+    
+    "environment": {
+        "platform": "aws",
+        "architecture": "amd64",
+        # ...others...
+        
+        # Configuration information is also included. 
+        "configuration": {
+            # The configuration id applied to the component
+            # before the test was run.
+            "component": "default",
+        }
+    },

```

Comment:
Environment is a little unusual in the results for every test, that seems like mostly a characteristic of a job and a lot of duplication?

Reply:
There are fixed aspects of the environment, like GCP vs AWS, but the enhancement suggests that environment include component configuration information -- a single job being able to apply and test multiple different configuration options. Imagine an operator being able to cycle through several of its typical configurations, running the same tests or tests specific to those configurations, during the execution of a single job. That configuration is relevant to the outcome of a test and Component Readiness must be able to differentiate the same test name running in one configuration vs another.

A next question would be why we would store those static environmental aspects in the aggregated results file alongside each test. My hope there is that the results files can begin to stand alone. You can just push the file content into a database and you know everything you need to know from the resulting DB. You don't need parse prowjob job names, for example, to derive additional context about how the test was run. Many tools can ingest a comprehensive file like this directly, so our options for analysis expand. Imagine wanting to move to a new database or use local tooling to analyze the data. With a comprehensive file format, we just ingest the file into our target analysis tool -- no custom logic like what we have in the cloud function required to pull bits and pieces from multiple artifacts.